selfhost/typechecker.jakt

```diff
@@ -5176,17 +5240,54 @@ struct Typechecker {
                     checked_arg = promoted_arg ?? checked_arg
 
                     // FIXME: It's super awkward that we have to pass "generic_inferences" here:
-                    .check_types_for_compat(lhs_type_id: param.variable.type_id, rhs_type_id: expression_type(checked_arg), generic_inferences: ["":""], span)
+                    .check_types_for_compat(lhs_type_id: param.variable.type_id, rhs_type_id: expression_type(checked_arg), generic_inferences: generic_substitutions, span)
 
                     args.push((call.name, checked_arg))
                 }
+
+                // We've now seen all the arguments and should be able to substitute the return type, if it's contains a
+                // type variable. For the moment, we'll just checked to see if it's a type variable.
+                if type_hint.has_value() {
+                    .check_types_for_compat(lhs_type_id: return_type, rhs_type_id: type_hint!, generic_inferences: generic_substitutions, span)
+                }
+                return_type = .substitute_typevars_in_type(type_id: return_type, generic_inferences: generic_substitutions)
```

Comment:
This line causes an infinite loop inside `substitute_typevars_in_type()`. I'm not sure why, as that seems to be a faithful recreation of the `substitute_typevars_in_type()` in the rust-based compiler, which works fine.